### DISCUSSION TITLE
Chaining / fluent-style syntax

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#14187] [Ideas]
EDIT: Turns out Google has already implemented this in all of their SQL variants (described at
https://research.google/pubs/sql-has-problems-we-can-fix-them-pipe-syntax-in-sql/) and a duckdb community member has implemented something similar (that'd need just a bit of polishing love) at https://github.com/ywelsch/duckdb-psql

Quoting from Google's paper:

> Many users understand pipe syntax immediately and find
it compelling, reporting that it greatly improves their productivity
and user experience in SQL. Several users have commented that
they “love” using SQL with pipe syntax. The pain points solved
here clearly resonate with users

---

It'd be great if 

```sql
<FIRSTQUERY>
|>
<SECONDQUERY>
|>
<THIRDQUERY>
```
(or something close to it) was translated to
```sql
FROM 
(
    FROM 
    (
       <FIRSTQUERY>
    )
    <SECONDQUERY>
)
<THIRDQUERY>
```

I recently found myself writing the latter a lot, where I edit-and-execute each level until I'm happy, then indent it, wrap it in parentheses, add a `FROM` above, and add more transformations below. This looks awful (although miles better than if `FROM`-first syntax didn't exist.)

I then started instead writing

```sql
WITH firstqueryname AS (
    <FIRSTQUERY>
), secondqueryname AS (
    FROM firstqueryname 
    <SECONDQUERY>
), thirdqueryname AS (
    FROM secondqueryname
    <THIRDQUERY>
)
FROM thirdqueryname
```
which is not much better for a few reasons:

1. You cannot continue writing at the bottom after `<FIRSTQUERY>` looks good. You instead have to indent it, wrap it in parentheses, and add the `WITH ` clause.
2. You have to name each sub-query. Often, this is unnatural and the only reason queries need to be broken up in the specific order is precedence (e.g., window functions being applied after `GROUP BY`) or aliasing rules; and if you try to give meaningful names to the sub-queries, you end up using longer and longer word chains. 
3. You have to type each sub-query name at least twice: once in the definition and once at the point of usage.
4. After every step, you have to edit the trailing `FROM <FINALQUERY>` (so really, you have to write each name three times), to complete the `WITH` clause and see your intermediate results.
5. When inserting a transformation step in the middle of the chain it's easy to forget to rename things so that the new step is actually used (and if you want to keep using meaningful names you might have to rename everything following in the chain as well).
6. Often, half of the lines you end up writing are just `<pickarbitraryname> AS ` and `FROM <pickarbitraryname>`.

(Most of these reasons are explained by the fact that `WITH` is strictly more powerful and can be used in situations where plain chaining is not appropriate. This is fine. I love `WITH`, too.)

I believe the syntax proposed in the introduction not only solves all these problems but also looks and lends itself to convenient iterative refinement of queries in the way that I at least usually use duckdb. A bonus advantage of a natural and clean way to break transformations into separate steps would be that it'd be less important to remember the precedence of `WHERE / HAVING / QUALIFY / LIMIT / USING SAMPLE / ....`  when you decided to cram everything into a single query to avoid a half dozen nested levels. 

For example, the proposed syntax would allow

```sql
FROM 
    df
|> -- Add lagged values
    SELECT
        Product,
        Date,
        CurrentValue,
        lag(CurrentValue, 1) OVER win AS Ago1,
        lag(CurrentValue, 2) OVER win AS Ago2,
    WINDOW win AS (PARTITION BY Product ORDER BY Date)
|> -- Compute correlations per product
    SELECT
        Product,
        corr(CurrentValue, Ago1) AS corr1,
        corr(CurrentValue, Ago2) AS corr2,
    GROUP BY ALL
|> -- Average correlations across products
    SELECT 
        mean(corr1) AS corr1,
        mean(corr2) AS corr2,
    GROUP BY ALL
|> -- Turn correlations into rows 
    UNPIVOT (
        correlation
        FOR lag IN (corr1, corr2)
    )
|> -- Skip trivial correlations
    WHERE 
        correlation < 1 AND correlation > -1
|> -- Extract lag as integer
    SELECT 
        regexp_extract(lag, 'corr(\d+)', 1)::INT AS lag,
        correlation
    ORDER BY
        lag
```

and you could imagine how before each `|>`, you could hit enter in your, say Jupyter notebook cell, to verify that what you get so far is what you expect, and then continue with the next steps without ever moving your cursor.

This would translate to

```sql
FROM (
    FROM (
        FROM (
            FROM (
                FROM (
                    FROM (
                        FROM 
                            df
                    )
                    SELECT
                        Product,
                        Date,
                        CurrentValue,
                        lag(CurrentValue, 1) OVER win AS Ago1,
                        lag(CurrentValue, 2) OVER win AS Ago2,
                    WINDOW win AS (PARTITION BY Product ORDER BY Date)
                )
                SELECT
                    Product,
                    corr(CurrentValue, Ago1) AS corr1,
                    corr(CurrentValue, Ago2) AS corr2,
                GROUP BY ALL
            )
            SELECT 
                mean(corr1) AS corr1,
                mean(corr2) AS corr2,
            GROUP BY ALL
        )
        UNPIVOT (
            correlation
            FOR lag IN (corr1, corr2)
        )
    )
    WHERE correlation < 1 AND correlation > -1
)
SELECT 
    regexp_extract(lag, 'corr(\d+)', 1)::INT AS lag,
    correlation
ORDER BY
    lag
```

and also be equivalent to

```sql
WITH raw AS (
    FROM 
        df
), with_lags AS (
    FROM 
        raw
    SELECT
        Product,
        Date,
        CurrentValue,
        lag(CurrentValue, 1) OVER win AS Ago1,
        lag(CurrentValue, 2) OVER win AS Ago2,
    WINDOW win AS (PARTITION BY Product ORDER BY Date)
), product_correlations AS (
    FROM 
        with_lags
    SELECT
        Product,
        corr(CurrentValue, Ago1) AS corr1,
        corr(CurrentValue, Ago2) AS corr2,
    GROUP BY ALL
), mean_correlations AS (
    FROM 
        product_correlations
    SELECT 
        mean(corr1) AS corr1,
        mean(corr2) AS corr2,
    GROUP BY ALL
), mean_correlations_unpivoted AS (
    FROM 
        mean_correlations
    UNPIVOT (
        correlation
        FOR lag IN (corr1, corr2)
    )
), mean_correlations_unpivoted_no_ones AS (
    FROM 
        mean_correlations_unpivoted
    WHERE 
        correlation < 1 AND correlation > -1
), mean_correlations_unpivoted_no_ones_renamed AS (
    FROM 
        mean_correlations_unpivoted_no_ones
    SELECT 
        regexp_extract(lag, 'corr(\d+)', 1)::INT AS lag,
        correlation
    ORDER BY
        lag
)
FROM mean_correlations_unpivoted_no_ones_renamed
```

The savings in the final number of lines (35 vs 40 and 50, respectively) and characters (879 vs 1105 and 1195, respectively) here is eclipsed by the superior readability and editability of the new syntax, the saving in mental effort to come up with throwaway names, and the back and forth editing effort while the pipeline is constructed layer by layer.

PS: You can try the two currently possible queries on 
```sql
CREATE OR REPLACE TABLE df AS (
    SELECT 1 AS Product, '2020-01-01'::DATE AS Date, 3 AS CurrentValue
    UNION
    SELECT 1 AS Product, '2020-01-02'::DATE AS Date, 4 AS CurrentValue
    UNION
    SELECT 1 AS Product, '2020-01-03'::DATE AS Date, 2 AS CurrentValue
    UNION
    SELECT 1 AS Product, '2020-01-04'::DATE AS Date, 2 AS CurrentValue
    UNION
    SELECT 2 AS Product, '2020-01-01'::DATE AS Date, 0 AS CurrentValue
    UNION
    SELECT 2 AS Product, '2020-01-02'::DATE AS Date, 2 AS CurrentValue
    UNION
    SELECT 2 AS Product, '2020-01-03'::DATE AS Date, 0 AS CurrentValue
    UNION
    SELECT 2 AS Product, '2020-01-04'::DATE AS Date, 1 AS CurrentValue
)
```

**Comment by soerenwolfers** [comment]
Turns out there is a community extension that does almost what I propose here: https://github.com/ywelsch/duckdb-psql

However, it seems more of an experiment and has a number of imperfections:
- It uses `|` for chaining, which can clash with the bitwise or operator
- it's based on raw string manipulation rather than hooking into the tokenization / parsing steps
- It rewrites to a CTE with autogenerated query aliases that could clash with user aliases (instead of translating to nested `FROM`s 
- Because it's based on raw string manipulation, it needs special syntax for chaining within subqueries

**Comment by ericemc3** [comment]
If find it very clearly explained, and why not? But i am not fond of pipes, it is realy painfull to type...
And a more fair comparaison would be with that actual syntax, we don't need those long CTE names. And often, i need to reuse 2 of them, for a join for instance, alias are then necessary.
```
WITH t1 AS ( -- Add lagged values
    FROM df
    SELECT
        Product,
        Date,
        CurrentValue,
        lag(CurrentValue, 1) OVER win AS Ago1,
        lag(CurrentValue, 2) OVER win AS Ago2,
    WINDOW win AS (PARTITION BY Product ORDER BY Date)
), t2 AS (   -- Compute correlations per product
    FROM t1
    SELECT
        Product,
        corr(CurrentValue, Ago1) AS corr1,
        corr(CurrentValue, Ago2) AS corr2,
    GROUP BY ALL
), t3 AS (  -- Average correlations across products
    FROM t2
    SELECT 
        mean(corr1) AS corr1,
        mean(corr2) AS corr2,
    GROUP BY ALL
), t4 AS ( -- Turn correlations into rows 
    FROM t3
    UNPIVOT (
        correlation
        FOR lag IN (corr1, corr2)
    )
) -- Extract lag as integer
FROM t5
SELECT 
    regexp_extract(lag, 'corr(\d+)', 1)::INT AS lag,
    correlation
WHERE 
	correlation < 1 AND correlation > -1
ORDER BY
    lag
```

**Reply by soerenwolfers** [reply]
On US keyboard layouts (I'm using one for programming and I'm not even an English native), the pipe character is easier than even a parenthesis. Anyway, not married to the character/keyword choice, if you think there is a better one. 

It wouldn't be difficult to allow optional aliases in a pipe syntax: Just do an equally trivial translation to CTE format, and use either an autogenerated alias or one given via

 ```sql
|> AS <ALIAS> SELECT ...
```

**Reply by ericemc3** [reply]
On my local - european - keyboard, |> requires 4 actions: Alt Gr + 6 then Shift + >

**Reply by soerenwolfers** [reply]
I mean, it's going to be hard to find a length-2 combination of special characters that does not require four key presses for *some* keyboard layout in the world. Also, the pipe character is all over the place in programming and everyday terminal use and `|>` is very common in other languages. Anyway, do tell if you know an alternative that doesn't clash with existing keywords or operators and still conveys the sense of piping.

**Reply by mskyttner** [reply]
Being a fan of the base R pipe operator |> ("and then", allowing to turn nesting inside out into a left-to-right sequence of steps), I feel this could offer a very nice way to untangle some of the complexity when composing nested sql interactively.... and although perhaps initially disruptive to some, perhaps it could be a quantum leap when moving towards friendlier and more intuitive SQL authoring.

Two different IDEs that I frequently use have a built-in key default binding (Ctrl-Shift-M for some reason) that makes that pipe operator easy to insert.

Optimally keyboards should have a separate key for it just next to the return button ;)

**Comment by soerenwolfers** [comment]
another community contribution that does essentially this (and more): https://wvlet.org/wvlet/docs/

**Comment by apstndb** [comment]
- Related: https://github.com/duckdb/duckdb/discussions/13599
